concept-art/has-field.sol

```diff
@@ -0,0 +1,52 @@
+data Unit = Unit
+data Pair(a, b) = Pair(a,b)
+
+type uint = word
+type string = word
+type bool = word
+
+data Memory(t) = Memory(Word)
+
+// this lets us link a given field in a struct to its position in it's
+// underlying generic representation as a tuple. 
+class self:Field(prevTypes, ty) {}
+
+// this struct should desugar into the following
+//struct S {
+//  f1 : uint;
+//  f2 : string;
+//  f3 : bool;
+//}
+
+// a type abstraction over tuples
+type s = Pair(uint, Pair(string, bool))
+
+// unique types identifying each field
+type sf1 = Unit
+type sf2 = Unit
+type sf3 = Unit
+
+// Field instances linking each field to it's position in the underlying tuple
+instance Pair(s, sf1):Field(Unit, uint) {}
+instance Pair(s, sf2):Field(Pair(uint, Unit), string) {}
+instance Pair(s, sf3):Field(Pair(uint, Pair(string, Unit)), bool) {}
+
+
+// struct field member access desugars into calls to this class
+class self:HasField(fieldType) {
+  function getField(x:sef) -> fieldType;
+}
+
+// we instantiate generic instances for references to types that implement Field
+instance (Pair(t, fieldName):Field(prevTypes, fieldType), fieldType:ValueType) => Pair(Memory(t), fieldName):HasField(fieldType) {
```

Comment:
I think it should probably actually be

```suggestion
instance (Pair(t, fieldName):Field(prevTypes, fieldType), fieldType:ValueType) => Pair(Memory(t), fieldName):HasField(memory(fieldType)) {
```

such that we can also assign *to* the thing in memory, not only read it (the implementation then just needs to bump the memory pointer of ``x.first`` by ``getMemorySize`` of ``prevTypes`` and we defer the actual load from memory.

Reply:
The signature of the ``getField`` function and its implementation would also need to change then, though :-).